substrate/client/network/src/litep2p/shim/notification/tests/fuzz.rs

```diff
@@ -0,0 +1,358 @@
+	let mut backed_off: HashMap<PeerId, Instant> = HashMap::new();
+
+	// The loop below is effectively synchronous, so for `PeerStore` & `ProtocolController`
+	// runners, spawned above, to advance, we use `spawn_blocking`.
+	let _ = tokio::task::spawn_blocking(move || {
+		// PRNG to use in `spawn_blocking` context.
+		let mut rng = rand::thread_rng();
+
+		// perform a certain number of actions while checking that the state is consistent.
+		//
+		// if we reach the end of the loop, the run has succeeded
+		for _ in 0..5000 {
+			// peer we are working with
+			let mut current_peer = None;
+
+			// current event for state transition validation
+			let mut current_event = None;
+
+			// last peer state for allowed event validation
+			let mut last_state = None;
+
+			// each of these weights corresponds to an action that we may perform
+			let action_weights = [150, 90, 90, 30, 30, 1, 1, 4, 4, 90];
+
+			match WeightedIndex::new(&action_weights).unwrap().sample(&mut rng) {
+				0 => match peerset.next().now_or_never() {
+					Some(Some(PeersetNotificationCommand::OpenSubstream { peers })) => {
+						for peer in peers {
+							assert!(opening.insert(peer));
+						}
+					},
+					Some(Some(PeersetNotificationCommand::CloseSubstream { peers })) => {
+						for peer in peers {
+							assert!(closing.insert(peer));
+						}
+					},
+					Some(None) => panic!("peerset exited"),
+					None => {},
+				},
+
+				// If we generate 1, discover a new node.
+				1 => {
+					let new_id = PeerId::random();
+					known_nodes.insert(new_id, State::Disconnected);
+					peer_store_handle.add_known_peer(new_id);
+				},
+
+				// If we generate 2, adjust a random reputation.
+				2 =>
+					if let Some(peer) = known_nodes.keys().choose(&mut rng) {
+						let val = Uniform::new_inclusive(i32::MIN, i32::MAX).sample(&mut rng);
+						peer_store_handle.report_peer(*peer, ReputationChange::new(val, ""));
+					},
+
+				// If we generate 3, disconnect from a random node.
+				3 =>
+					if let Some(peer) = connected_nodes.iter().choose(&mut rng).cloned() {
+						log::info!("Disconnected from {}", peer);
+						connected_nodes.remove(&peer);
+
+						let state = known_nodes.get_mut(&peer).unwrap();
+						last_state = Some(*state);
+						*state = State::Disconnected;
+
+						peerset.report_substream_closed(peer);
+
+						current_peer = Some(peer);
+						current_event = Some(Event::Disconnected);
+					},
+
+				// If we generate 4, get an inbound connection from a random node
+				4 => {
+					if let Some(peer) = known_nodes
+						.keys()
+						.filter(|n| opening.iter().all(|m| m != *n) && !connected.contains(*n))
+						.choose(&mut rng)
+						.cloned()
+					{
+						log::info!("Incoming connection from {peer}, index {next_incoming_id}");
+
+						match peerset.report_inbound_substream(peer) {
+							ValidationResult::Accept => {
+								opening.insert(peer);
+							},
+							ValidationResult::Reject => {},
+						}
+					}
```

Comment:
Is it intended that we don't register `current_peer`, `current_event`, and don't update `*state` in this branch? I see it's also not updated in `9`. The events are not checked against `allowed_events` either.

And, if incoming id is actually not needed for the new `Peerset`, we can get rid of `State`/`BareState` enum separation.

Reply:
I've rewritten the fuzz test to better match this design of `Peerset`. `current_peer`, `current_event` and `state` don't make sense IMO because `Peerset` and litep2p communicate over an asynchronous link. Please take a look.

Reply:
You are right, `current_peer`, `current_event` and `state` don't make sense anymore in the original form as we are handling more than one peer on every iteration of the loop due to batching. It is possible to make use of something similar (i.e., keeping track of peer states in a `HashMap` and validating the transitions) anyway, but you are doing essentially the same with separate `opening`/`open`/`closing`/`closed`. For me, it's easier to reason about the peer state kept in one place, but your approach is also fine. I'm going to resolve this.